Apt generator

velo opened the pull request.

velo committed:
Initial hand made implementation for feign client

velo commented on apt-generator/src/test/java/feign/aptgenerator/github/TypeReference.java:
```diff
@@ -0,0 +1,70 @@
+ * which can be passed to methods that accept TypeReference, or resolved using
+ * <code>TypeFactory</code> to obtain {@link ResolvedType}.
+ */
+public abstract class TypeReference<T> implements Comparable<TypeReference<T>> {
```
this need to be deleted

velo commented on apt-generator/src/test/java/feign/aptgenerator/github/GitHubFactoryExample.java:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  public static void main(String... args) {
+    final GitHub github = new GitHubFactory(config());
```
Example on how the generated code would be invoked

velo commented on apt-generator/src/test/java/feign/aptgenerator/github/GitHubFactory.java:
```diff
@@ -0,0 +1,104 @@
+import feign.Target.HardCodedTarget;
+import feign.example.github.GitHubExample.GitHub;
+
+public class GitHubFactory implements GitHub {
```
once the generator is implemented, this is the code it would generate

SchulteDev commented on core/pom.xml:
```diff
@@ -32,6 +32,11 @@
   <dependencies>
+    <dependency>
+      <groupId>org.projectlombok</groupId>
+      <artifactId>lombok</artifactId>
```
Can be provided, see https://projectlombok.org/setup/maven

velo replied:
"provided" is declared on the parent pom

SchulteDev commented:
Thanks for taking care, @velo !
I do not really figure out the important part in this pull request. You are introducing a new module "apt-generator". How would it be used?
As far as I understand, this PR encapsulates more than one relevant change. In my opinion, it should be split. For example, one PR only with the code cleanup (changes to import-statements, finals, visibility of methods, ...). Another PR with the changes to FeignConfig.builder() (if that is possible). Another PR with the relevant changes for the new module.
I am really thankful for your work, and I don't want to cause you more work, but I think it is important to focus in each PR on a specific task.

velo commented:
@SchulteMarkus 

> "apt-generator". How would it be used?

once implemented (right now is just an example of what the generator would produce)
My plan is to have an APT generator that can be executed using maven, gradle, sbt or javac
https://www.baeldung.com/java-annotation-processing-builder

velo commented:
BTW, can you help me on automating this to run using graal VM?

SchulteDev commented:
Ok, I see.

I think, a staring point is to test, if GaalVM-compiler compiles this Java, too.

     feign (apt-generator)$ docker run --rm -v $(pwd):/feign oracle/graalvm-ce:1.0.0-rc16 \
      sh -c "(cd /feign && ./mvnw package -T1C -DskipTests)"

It seems to do so.

For testing, if GaalVM native-image generation actual works with feign as a dependency, we would have to have a concrete application with a Java main class. Hmm, I don't know where to start from here.
The actual build of a native-image should not be that complex, once we have a test application, see https://medium.com/graalvm/simplifying-native-image-generation-with-maven-plugin-and-embeddable-configuration-d5b283b92f57

velo committed:
Create a test that compile native using graalVM

velo commented:
@SchulteMarkus seems this code generation approach would work on graal.

I still wanna get ride of TypeReference

snyk-bot committed:
fix: httpclient/pom.xml to reduce vulnerabilities (#997)

The following vulnerabilities are fixed with an upgrade:
- https://snyk.io/vuln/SNYK-JAVA-ORGAPACHEHTTPCOMPONENTS-31517

velo committed:
Reusable contracts

velo committed:
Declarative contracts

velo committed:
APT generator generating MethodMetadata

velo committed:
Remove Arrays import

velo committed:
Actually using the data structure to read declaritve contracts

velo committed:
Merge branch 'declarative_contracts' into apt-generator

velo committed:
Merge remote-tracking branch 'upstream/master' into apt-generator

velo committed:
Merge remote-tracking branch 'upstream/master' into apt-generator

velo committed:
wip
